### ISSUE TITLE
Propogate accent color changes to linear gradients? (1.6.5)

### ISSUE THREAD

**Issue by MV10** [MahApps/MahApps.Metro#3463]
Edit: I changed the title to be specific to the linear gradient; as luck would have it, that was the first place I tried to use a MahApps accent, but it works as expected literally everywhere else in the app.

I can't seem to figure out how to propogate an accent-change to user controls that are styled with MahApps accent and theme colors. I am using `DynamicResource` and the change event, and I've tried several approaches to the `ChangeAppTheme` call (app, window, various `Resources` references, etc.)

It's easy to reproduce:

#### Merged in App.xaml
```xml
<ResourceDictionary Source="pack://application:,,,/MahApps.Metro;component/Styles/Controls.xaml" />
<ResourceDictionary Source="pack://application:,,,/MahApps.Metro;component/Styles/Fonts.xaml" />
<ResourceDictionary Source="pack://application:,,,/MahApps.Metro;component/Styles/Colors.xaml" />
<ResourceDictionary Source="pack://application:,,,/MahApps.Metro;component/Styles/Accents/Cobalt.xaml" />
<ResourceDictionary Source="pack://application:,,,/MahApps.Metro;component/Styles/Accents/BaseLight.xaml" />
<ResourceDictionary Source="WatermarkBrush.xaml"/>
<ResourceDictionary Source="WatermarkStyle.xaml"/>
```

#### UC-related XAMLs
```xml
<LinearGradientBrush x:Key="WatermarkBrush" StartPoint="0.5,0" EndPoint="0.5,1">
    <GradientStop Color="{DynamicResource AccentColor}" Offset="0.15"/>
    <GradientStop Color="{DynamicResource AccentColor4}" Offset="1"/>
</LinearGradientBrush>

<Style x:Key="WatermarkStyle" TargetType="Label">
    <Style.Setters>
        <Setter Property="Foreground" Value="{DynamicResource WatermarkBrush}"/>
        <Setter Property="FontFamily" Value="Arial Black"/>
        <Setter Property="FontSize" Value="36"/>
        <Setter Property="FontWeight" Value="Bold"/>
    </Style.Setters>
</Style>

<StackPanel>
    <Label Content="MyApplicationName" Style="{DynamicResource WatermarkStyle}"/>
</StackPanel>
```

#### MainWindow content
```xml
<StackPanel>
    <Button Content="Blue on White" HorizontalAlignment="Center" VerticalAlignment="Center" Margin="0,5" Click="BlueOnWhite"/>
    <Button Content="Red on Black" HorizontalAlignment="Center" VerticalAlignment="Center" Margin="0,5" Click="RedOnBlack"/>
    <local:Watermark x:Name="MyUserControl" HorizontalAlignment="Center" Margin="30"/>
</StackPanel>
```

#### Code
```csharp
public MainWindow()
{
    InitializeComponent();

    //ThemeManager.IsThemeChanged += (s, args) =>
    //    ThemeManager.ChangeAppStyle(Application.Current, args.Accent, args.AppTheme);

    //ThemeManager.IsThemeChanged += (s, args) =>
    //    ThemeManager.ChangeAppStyle(this, args.Accent, args.AppTheme);

    ThemeManager.IsThemeChanged += (s, args) =>
        ThemeManager.ChangeAppStyle(MyUserControl.Resources, args.Accent, args.AppTheme);
}

private void BlueOnWhite(object sender, RoutedEventArgs args)
    => ThemeManager.ChangeAppStyle(Application.Current, ThemeManager.GetAccent("Cobalt"), ThemeManager.GetAppTheme("BaseLight"));

private void RedOnBlack(object sender, RoutedEventArgs args)
    => ThemeManager.ChangeAppStyle(Application.Current, ThemeManager.GetAccent("Crimson"), ThemeManager.GetAppTheme("BaseDark"));
```

I've tried _many_ more variations on `ChangeAppStyle` than what is shown above. The accents defined in App.xaml are applied at startup, but no matter how I respond to the theme-change event, the changes are not reflected in the UC. I even tried removing and re-adding the UC but it retains the startup style. (I also suspect my style and brush XAML should use `StaticResource` and only the UC needs `DynamicResource`, but obviously I'm a bit unclear about how the resource-change process works.) 

I'm sure it's something simple. What am I missing?

![image](https://user-images.githubusercontent.com/794270/53285653-04810400-3731-11e9-851d-320643ec1851.png)

**Comment by MV10** [comment]
Perhaps this problem is specific to the `LinearGradientBrush` or maybe how I reference the accent color there? I added a `Border` with an accent-dependent style and that works as expected. I don't even need the event, `DynamicResource` is enough.

```xml
<Style x:Key="BorderStyle" TargetType="Border">
    <Style.Setters>
        <Setter Property="CornerRadius" Value="10"/>
        <Setter Property="BorderThickness" Value="5"/>
        <Setter Property="BorderBrush" Value="{DynamicResource AccentColorBrush}"/>
    </Style.Setters>
</Style>

<StackPanel>
    <Border Margin="50,10" Padding="5" Style="{DynamicResource BorderStyle}">
        <Label Content="MyApplicationName" Style="{DynamicResource WatermarkStyle}"/>
    </Border>
</StackPanel>
```

![image](https://user-images.githubusercontent.com/794270/53286430-7874d980-373c-11e9-85cc-b800d3017652.png)

**Comment by timunie** [comment]
I have the same problem. If i save my accentcolor and reopen my app the brush works as expected. As a work around i put all my styles and brushes into a new resourcedictionary and reload the dictionary everytime i change the accentcolor. 

Is there a better way? 

@punker76  How do you update the dynamic resources when you change the Theme? Is there any trick?

**Comment by punker76** [comment]
@MV10 I created a short sample with your source snippets and I can't reproduce this issue. Also you don't need to change the app style again at the ThemeManager.IsThemeChanged event.

/cc @timunie 

repo: https://github.com/punker76/MahApps.Metro.LinearGradientBrush.Theme.Sample

![mahapps_theme_lineargradientbrush](https://user-images.githubusercontent.com/658431/58588674-77ce7080-8260-11e9-88e0-480331d8be84.gif)

**Window**

```xaml
<mah:MetroWindow x:Class="WpfApp3.MainWindow"
                 xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
                 xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
                 xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
                 xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
                 xmlns:mah="http://metro.mahapps.com/winfx/xaml/controls"
                 xmlns:wpfApp3="clr-namespace:WpfApp3"
                 mc:Ignorable="d"
                 Title="MainWindow" Height="450" Width="800">
  <Grid>

    <StackPanel VerticalAlignment="Center" HorizontalAlignment="Center">
      <Button Content="Blue on White" HorizontalAlignment="Center" VerticalAlignment="Center" Margin="0,5"
              Click="BlueOnWhite" />
      <Button Content="Red on Black" HorizontalAlignment="Center" VerticalAlignment="Center" Margin="0,5"
              Click="RedOnBlack" />
      <wpfApp3:TestUserControl />
    </StackPanel>

  </Grid>
</mah:MetroWindow>
```

```csharp
using System.Windows;
using MahApps.Metro;
using MahApps.Metro.Controls;

namespace WpfApp3
{
  /// <summary>
  /// Interaction logic for MainWindow.xaml
  /// </summary>
  public partial class MainWindow : MetroWindow
  {
    public MainWindow()
    {
      InitializeComponent();
    }

    private void BlueOnWhite(object sender, RoutedEventArgs args)
      => ThemeManager.ChangeAppStyle(Application.Current, ThemeManager.GetAccent("Cobalt"), ThemeManager.GetAppTheme("BaseLight"));

    private void RedOnBlack(object sender, RoutedEventArgs args)
      => ThemeManager.ChangeAppStyle(Application.Current, ThemeManager.GetAccent("Crimson"), ThemeManager.GetAppTheme("BaseDark"));
  }
}
```

**UC**

```xaml
<UserControl x:Class="WpfApp3.TestUserControl"
             xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             mc:Ignorable="d"
             d:DesignHeight="450" d:DesignWidth="800">
  <Grid>

    <Grid.Resources>
      <LinearGradientBrush x:Key="WatermarkBrush" StartPoint="0.5,0" EndPoint="0.5,1">
        <GradientStop Color="{DynamicResource AccentColor}" Offset="0.15" />
        <GradientStop Color="{DynamicResource AccentColor4}" Offset="1" />
      </LinearGradientBrush>

      <Style x:Key="WatermarkStyle" TargetType="Label">
        <Style.Setters>
          <Setter Property="Foreground" Value="{DynamicResource WatermarkBrush}" />
          <Setter Property="FontFamily" Value="Arial Black" />
          <Setter Property="FontSize" Value="36" />
          <Setter Property="FontWeight" Value="Bold" />
        </Style.Setters>
      </Style>

      <Style x:Key="BorderStyle" TargetType="Border">
        <Style.Setters>
          <Setter Property="CornerRadius" Value="10" />
          <Setter Property="BorderThickness" Value="5" />
          <Setter Property="BorderBrush" Value="{DynamicResource AccentColorBrush}" />
        </Style.Setters>
      </Style>
    </Grid.Resources>

    <StackPanel VerticalAlignment="Center" HorizontalAlignment="Center">
      <Border Margin="50,10" Padding="5" Style="{DynamicResource BorderStyle}">
        <Label Content="MyApplicationName" Style="{DynamicResource WatermarkStyle}" />
      </Border>
    </StackPanel>

  </Grid>
</UserControl>
```

**Comment by timunie** [comment]
Thank you for your feedback. @punker76 i provided a colorpicker for the user in my app and update all the accentbrushes.  This works for all solidcolorbrushes but not for gradientbrushes. I will try it again in a sample app and let you know if it works now. It would be cool 😎 if someday the ThemeManager accepts cutom colors. 

Have a nice evening

**Comment by timunie** [comment]
I tried it again and it worked as expected, but it does not work if I just change the AccentColors. If I just change the AccentColors it won't work. I had to create a new `Theme`. 

I extended the ThemeManager to create a custom Theme (light and dark) based on userinput (I found a similar code in an example application :-)). I made a pull request (#3515) for my changes, Maybe it will be merged some day and others could benefit of this feature, I am waiting for the feedback of @punker76

Happy coding
